### ISSUE TITLE
Mixed MySQL string collations in different environments/deployments

### ISSUE THREAD

**Issue by ysbaddaden** [instedd/cdx#1517]
In light of #1498 we noticed that there are different string collations in the different MySQL databases of CDx. For example for local development @straight-shoota and I have `utf8_unicode_ci` but in staging we noticed it's `latin1_swedish_ci`? We don't about the production deployments.

We should investigate and have a single collation for all environments that we can rely on. Ideally it should be `utf8mb4_unicode_ci` which is the current recommendation.

For reference, this is what we used to notice the difference:
```sql
SELECT COLUMN_NAME, COLLATION_NAME
FROM information_schema.columns
WHERE TABLE_NAME = 'batches' AND COLUMN_NAME IN ('isolate_name', 'batch_number');
```

**Comment by ysbaddaden** [comment]
@matiasgarciaisaia when you have some minutes :)

**Comment by matiasgarciaisaia** [comment]
Staging:

```
mysql> SELECT COLUMN_NAME, COLLATION_NAME
    -> FROM information_schema.columns
    -> WHERE TABLE_NAME = 'batches' AND COLUMN_NAME IN ('isolate_name', 'batch_number');
+--------------+-------------------+
| COLUMN_NAME  | COLLATION_NAME    |
+--------------+-------------------+
| isolate_name | latin1_swedish_ci |
| batch_number | latin1_swedish_ci |
+--------------+-------------------+
2 rows in set (0.03 sec)
```

Production:

```
mysql> SELECT COLUMN_NAME, COLLATION_NAME
    -> FROM information_schema.columns
    -> WHERE TABLE_NAME = 'batches' AND COLUMN_NAME IN ('isolate_name', 'batch_number');
+--------------+-----------------+
| COLUMN_NAME  | COLLATION_NAME  |
+--------------+-----------------+
| isolate_name | utf8_unicode_ci |
| batch_number | utf8_unicode_ci |
+--------------+-----------------+
2 rows in set (0.00 sec)
```

RadX:

```
mysql> SELECT COLUMN_NAME, COLLATION_NAME
    -> FROM information_schema.columns
    -> WHERE TABLE_NAME = 'batches' AND COLUMN_NAME IN ('isolate_name', 'batch_number');
+--------------+--------------------+
| COLUMN_NAME  | COLLATION_NAME     |
+--------------+--------------------+
| isolate_name | utf8mb4_unicode_ci |
| batch_number | utf8mb4_unicode_ci |
+--------------+--------------------+
2 rows in set (0.01 sec)
```

![Where we're going we don't need consistency](https://user-images.githubusercontent.com/5470539/157975880-d9ddf884-3e69-4f42-a484-0cef3a759f56.png)

**Comment by ysbaddaden** [comment]
It appears that `latin1_swedish_ci` was the default collation for the `latin1` charset.

I thought maybe `config/database.yml` was at fault, but it always used `encoding: utf8` (or `unicode`) so that's probably not it.

The databases were created from different MySQL servers, and they had different defaults or configuration? First, staging with the `latin1` charset (:fearful:), then `production` with `utf8` (better) and recently radx with the now recommended `utf8mb4` charset (which is _finally_ proper UTF-8).

@straight-shoota a migration to harmonize on `utf8mb4` in every table would be a good idea.

**Comment by straight-shoota** [comment]
MySQL is super weird. They failed to implement UTF-8 properly and kept `latin1` as default for years, while Unicode was long the standard everywhere else. I believe the (fixed) UTF-8 implementation `utf8mb` only became the default character set recently (in 8.0).

> @straight-shoota a migration to harmonize on utf8mb4 in every table would be a good idea.

Yes. I don't know how this works in MySQL, but I'm sure there are some guides/blog posts. Just need to figure out which ones are actually good 🤷 
Can we get a dump of the production database for testing this off-line first?

**Comment by ysbaddaden** [comment]
We should use the staging database for tryouts, being `latin1` it's the most likely to fail hideously, especially with ActiveRecord setting the connection to use `utf8` (:fearful:). The `utf8` to `utf8mb4` should be more straightforward (but maybe I'm overconfident).

**Comment by matiasgarciaisaia** [comment]
I second @ysbaddaden's approach: let's write _whatever is needed_ testing with staging, then we try that offline with production as suggested by @straight-shoota 👍

My fear is that this is not something you can change on the fly. But I really don't know, so I hope I'm wrong on this one.

Even if we have to drop the DB and recreate it, it's probably not that big - so we can afford a really short maintenance window.

----

I think the disparity comes from the fact that we first manually ask the DB server to create the DB, then we late ActiveRecord automatically create the tables, schemas, etc. So by the time ActiveRecord kicks in, we've already messed stuff up. For production, it has 5+ years - I'm not even sure if we had already converged on the proper UTF-8 type back then. For staging, it was me who set it up kind of recently, and I guess I didn't pay enough attention then. Weird thing is I did pay attention to this while creating the production database 🤷

**Comment by straight-shoota** [comment]
This should really work without rebuilding the db. What we need to test for is how much time it takes.
I wouldn't expect much actually, because `utf8mb` should be binary compatible with `utf8` and `latin1` for the basic latin block. And there are probably not many codepoints outside that block (if at all). So assuming MySQL is clever about it, there should be little to no changes to actual data.
For testing purposes we should make sure to use some codepoints outside Basic Latin before migrating the charset.

**Comment by ysbaddaden** [comment]
Some pointers as to what to do:
https://stackoverflow.com/questions/38363566/trouble-with-utf-8-characters-what-i-see-is-not-what-i-stored

**Comment by ysbaddaden** [comment]
The one thing we need to know is how are non ASCII characters such as `é` stored in the table: as `latin1` (1 byte) or `utf8` (2 bytes). We do have `SET NAMES utf8` thanks to ActiveRecord, but we don't know about.

Note: `utf8` or `utf8mb4` seem to be equivalent, the only difference being support for 4 characters, which are needed for Emojis (we don't care) but also some Chinese characters (we may care).

**Comment by matiasgarciaisaia** [comment]
I've just confirmed that with the new setting changed in be279aafb1d506218eedc45ead3cb0a934a0f7bb you can use Japanese characters (`揺り動かす`) and emojis (`👨‍👨‍👧‍👧` & `🙃`) in the user's & institution's names.

I still have to check how to migrate the DBs with the old encodings to the new ones (and how Spanish accents, for example, migrate there).

**Comment by straight-shoota** [comment]
Note that `encoding` in `config/database.yml` (https://github.com/instedd/cdx/commit/be279aafb1d506218eedc45ead3cb0a934a0f7bb) only affects the encoding settings used for connections to the database, not the database itself.

**Comment by matiasgarciaisaia** [comment]
Yes! Thanks for double checking :)

I intend to come up with a DB migration that can change every DB we have to the new encoding. I'm not sure if that's technically feasible, but that's what I'm aiming for.

**Comment by matiasgarciaisaia** [comment]
```sql
SELECT table_name, COLUMN_NAME, COLLATION_NAME
FROM information_schema.columns
where table_schema = 'cdp' and collation_name is not null and collation_name != 'utf8mb4_unicode_520_ci';
```



table_name | COLUMN_NAME | COLLATION_NAME
-- | -- | --
ar_internal_metadata | key | latin1_swedish_ci
ar_internal_metadata | value | latin1_swedish_ci
boxes | uuid | latin1_swedish_ci
boxes | site_prefix | latin1_swedish_ci
boxes | core_fields | latin1_swedish_ci
boxes | custom_fields | latin1_swedish_ci
boxes | purpose | latin1_swedish_ci
schema_migrations | version | latin1_swedish_ci

I guess I'm OK with the `ar_internal_metadata` and `schema_migrations` tables using other encoding (it's Rails internal stuff, I guess they won't be including emojis there), but the `boxes` one is a PR race-condition?

**Comment by straight-shoota** [comment]
Yeah, apparently both migrations where applied at the same time, and then the order matters. The boxes migration ran before the charset migration and the new table wasn't covered.
That's the kind of thing why I would've used `SHOW TABLES` instead of a fixed list.
